wrappers/kmip/kmip.go

```diff
@@ -0,0 +1,537 @@
+		))
+	}
+
+	ciphers := []string{}
+	if !opts.Options.WithDisallowEnvVars {
+		cipherStr := os.Getenv(EnvKmipTls12Ciphers)
+		if cipherStr != "" {
+			ciphers = strings.Split(cipherStr, ",")
+		}
+	}
+	if len(ciphers) == 0 {
+		ciphers = opts.withTls12Ciphers
+	}
+	if len(ciphers) > 0 {
+		k.opts = append(k.opts, kmipclient.WithTlsCipherSuiteNames(ciphers...))
+	}
+
+	cpname := ""
+	if !opts.Options.WithDisallowEnvVars {
+		cpname = os.Getenv(EnvKmipEncryptAlg)
+	}
+	if cpname == "" {
+		cpname = opts.withCryptoParams
+	}
+	var ok bool
+	k.cryptoParams, ok = cryptoParamsPreset[cpname]
+	if !ok {
+		return nil, fmt.Errorf("invalid crypto parameters %q", cpname)
+	}
+	mId := slices.Index(mechanisms, cpname)
+	if mId < 0 {
+		return nil, fmt.Errorf("KMIP cryptographic parameter does not have an associated mechanism ID")
+	}
+	k.mechanismID = uint64(mId)
+
+	// Encrypt / Decrypt operations support has been added in kmip v1.2. So we cannot support v1.0 nor v1.1
+	k.opts = append(k.opts, kmipclient.WithKmipVersions(kmip.V1_2, kmip.V1_3, kmip.V1_4))
+
+	conn, err := k.connect(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial kmip connection endpoint: %w", err)
+	}
+	defer conn.Close()
+
+	if err := k.verifyKey(ctx, conn); err != nil {
+		return nil, fmt.Errorf("invalid kmip key object: %w", err)
+	}
+
+	// Cache client so we can clone it to reopen it on demand, skipping protocol version negociation
+	k.cachedClient = conn
+
+	// Map that holds non-sensitive configuration info
+	wrapConfig := new(wrapping.WrapperConfig)
+	wrapConfig.Metadata = make(map[string]string)
+	wrapConfig.Metadata["kms_key_id"] = k.keyId
+	wrapConfig.Metadata["endpoint"] = k.endpoint
+	wrapConfig.Metadata["timeout"] = strconv.Itoa(int(k.timeout))
+	wrapConfig.Metadata["encrypt_alg"] = cpname
+	wrapConfig.Metadata["kmip_version"] = conn.Version().String()
+	if serverName != "" {
+		wrapConfig.Metadata["server_name"] = serverName
+	}
+	if len(ciphers) != 0 {
+		wrapConfig.Metadata["kmip_tls12_ciphers"] = strings.Join(ciphers, ",")
+	}
+	if k.pubKeyId != "" {
+		wrapConfig.Metadata["kms_public_key_id"] = k.pubKeyId
+	}
+
+	return wrapConfig, nil
+}
+
+// Type returns the type for this particular wrapper implementation
+func (k *Wrapper) Type(_ context.Context) (wrapping.WrapperType, error) {
+	return wrapping.WrapperTypeKmip, nil
+}
+
+// KeyId returns the last known key id
+func (k *Wrapper) KeyId(_ context.Context) (string, error) {
+	return k.keyId, nil
+}
+
+// Encrypt is used to encrypt the master key using the the KMIP.
+// This returns the ciphertext, and/or any errors from this
+// call. This should be called after the KMS client has been instantiated.
+func (k *Wrapper) Encrypt(ctx context.Context, plaintext []byte, opt ...wrapping.Option) (*wrapping.BlobInfo, error) {
+	if plaintext == nil {
+		return nil, fmt.Errorf("given plaintext for encryption is nil")
+	}
+
+	env, err := wrapping.EnvelopeEncrypt(plaintext, opt...)
```

Comment:
Like with #14, I wouldn't use envelope encryption. As we consider whether to make this more general-purpose (and support PKCS#11/KMIP/KMS backed keys in secret engines, we may or may not use g-k-w and so want to do the simplest thing here). 

OpenBao for auto-unseal just uses this to encrypt a small AES-256 key so this should never hit RSA limits unless using tiny key sizes.

Reply:
Makes sense, I removed envelope encryption